### PR TITLE
Fix unsanitized nulls output from percentile_approx with partial empty input

### LINKED ISSUE #24056: [BUG] percentile_approx returns non-empty null list rows for mixed empty and non-empty TDigests
**Describe the bug**

`cudf::percentile_approx` can return a LIST column with non-empty null rows when its
TDigest input mixes empty and non-empty digests. The empty digest rows are marked null,
but their list offsets still reserve `percentiles.size()` child elements. This violates
the nested-column invariant expected by downstream consumers; the Java API reports:

```text
Column has non-empty nulls
```

This is related to #10856 / #11498, which sanitize the all-empty-input case. The mixed
case still takes the regular output path and remains unsanitized.

The problem was originally observed by cuDF for Apache Spark in
NVIDIA/cudf-spark#14634.

**Steps/Code to reproduce bug**

The following libcudf gtest setup reproduces the mixed case on the validated `main`
base (`2cf22a1fb5b864bd53128034d34217943bea0047`):

```cpp
auto const values =
  cudf::test::fixed_width_column_wrapper<double>{{1, 0, 3}, {true, false, true}};
auto const keys        = cudf::test::fixed_width_column_wrapper<int32_t>{0, 1, 2};
auto const percentiles = cudf::test::fixed_width_column_wrapper<double>{0.0, 0.5, 1.0};

cudf::groupby::groupby gb(
  cudf::table_view{{keys}}, cudf::null_policy::EXCLUDE, cudf::sorted::YES);
std::vector<cudf::groupby::aggregation_request> requests;
std::vector<std::unique_ptr<cudf::groupby_aggregation>> aggregations;
aggregations.push_back(cudf::make_tdigest_aggregation<cudf::groupby_aggregation>(1000));
requests.push_back({values, std::move(aggregations)});
auto const tdigest_column = gb.aggregate(requests);

cudf::tdigest::tdigest_column_view tdv(*tdigest_column.second[0].results[0]);
auto const result = cudf::percentile_approx(tdv, percentiles);
```

An equivalent pre-fix native regression run failed with:

```text
C++ exception with description "lhs column has non-empty nulls" thrown in the test body.
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test
```

The regular path currently creates fixed-width offsets for every digest, nullifies empty
digest rows, and leaves the child slots for those null rows in place. The percentile
kernel does not populate those slots.

**Expected behavior**

For the input above, the result should have:

```text
offsets:  [0, 3, 3, 6]
child:    [1, 1, 1, 3, 3, 3]
validity: [true, false, true]
```

In other words, an empty digest should produce a null list row with zero child elements.

A local candidate fix (`0ee212f124a9970cffb9a99c13766f58e7bed429`) gives empty digests zero-length list ranges and writes non-empty
digests directly into a compact child column. The no-empty kernel path remains dense and
does not load output offsets. Validation includes a nullable percentile and mixed-input
validity-mask boundaries at 31, 32, 33, 65, and 257 rows:

```text
[  PASSED  ] 2 tests.  # focused mixed empty/non-empty regressions
[==========] 388 tests from 67 test suites ran.
[  PASSED  ] 388 tests.
```

**Environment overview (please complete the following information)**

- Environment location: bare metal
- Method of cuDF install: built from source
- Build host: NVIDIA A100, CUDA 12.9.1
- Runtime GPU: NVIDIA RTX 5880 Ada Generation
- CUDA architectures: `80-real;89-real`

**Environment details**

The failure and passing validation above are native libcudf `QUANTILES_TEST` runs from
the stated source commit. A full Spark/JNI validation requires rebuilding the combined
spark-rapids-jni native library; swapping only standalone `libcudf.so` into an existing
JNI artifact is not ABI-complete.

**Additional context**

The proposed change is localized to `cpp/src/quantiles/tdigest/tdigest.cu`. For mixed
input, build list offsets with row size zero for empty digests, allocate only the compact
child size, and write valid digest results through those offsets. This avoids an
additional full-column gather and its temporary output allocation. The no-empty fast
path is unchanged. The percentile kernel uses a bounded launch and a grid-stride loop so
a sparse compact output cannot request an invalid CUDA grid from the larger dense pair
count.

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fix `percentile_approx` for mixed empty and non-empty TDigest inputs by giving empty output rows zero-length list ranges and compacting the child values. Also adds a regression test and preserves the existing dense fast path for inputs without empty digests.

Closes https://github.com/NVIDIA/cudf/issues/24056

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 1]
```diff
@@ -59,102 +59,106 @@ struct make_centroid {
 // kernel for computing percentiles on input tdigest (mean, weight) centroid data.
 template <typename CentroidIter>
 CUDF_KERNEL void compute_percentiles_kernel(device_span<int32_t const> tdigest_offsets,
+                                            device_span<int32_t const> output_offsets,
+                                            bool output_is_dense,
```
Comment (davidwendt) [inline cpp/src/quantiles/tdigest/tdigest.cu:65]: Recommend using "Hide whitespace" when reviewing this section.

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 2]
```diff
@@ -346,32 +359,7 @@ std::unique_ptr<column> percentile_approx(tdigest_column_view const& input,
...
-      input.size(),
-      std::move(offsets),
-      cudf::make_empty_column(type_id::FLOAT64),
-      input.size(),
-      cudf::detail::create_null_mask(
-        input.size(), mask_state::ALL_NULL, cuda::stream_ref(stream), mr));
-  }
-
-  // if any of the input digests are empty, nullify the corresponding output rows (values will be
-  // uninitialized)
```
Comment (davidwendt) [inline cpp/src/quantiles/tdigest/tdigest.cu:374]: The above code is mostly just moved below the `bitmask, null_count` statement.

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 3]
```diff
@@ -389,11 +377,39 @@ std::unique_ptr<column> percentile_approx(tdigest_column_view const& input,
...
+    data_type{type_id::INT32}, input.size() + 1, mask_state::UNALLOCATED, stream, mr);
+  auto const tdigest_sizes = detail::size_begin(tdv);
+  auto row_size_iter       = cudf::detail::make_counting_transform_iterator(
+    0,
+    cuda::proclaim_return_type<size_type>(
+      [tdigest_sizes, num_rows = input.size(), num_percentiles = percentiles.size()] __device__(
+        size_type row) -> size_type {
+        if (row == num_rows) { return 0; }
+        return tdigest_sizes[row] == 0 ? 0 : num_percentiles;
+      }));
```
Comment (davidwendt) [inline cpp/src/quantiles/tdigest/tdigest.cu:396]: This fancy iterator helps ensure we do not read past the end of the input memory.

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 4]
```diff
@@ -389,11 +377,39 @@ std::unique_ptr<column> percentile_approx(tdigest_column_view const& input,
...
+  }
+
+  auto const output_offsets = device_span<int32_t const>{offsets->view().begin<int32_t>(),
+                                                         static_cast<size_t>(offsets->size())};
+  return cudf::make_lists_column(
+    input.size(),
+    std::move(offsets),
+    detail::compute_approx_percentiles(input, percentiles, output_offsets, null_count, stream, mr),
+    null_count,
+    std::move(bitmask));
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/quantiles/tdigest/tdigest.cu:419]: `make_lists_column` accepts only  `(num_rows, offsets_column, child_column, null_count, null_mask)`
It has no stream or mr parameters, because it only adopts already-created columns and mask buffers; those are already allocated/generated with the supplied stream and mr in this function.

**Comment by wjxiz1992** [comment]
I was independently working from #24056 and opened #24103 before noticing this PR. I compared the exact heads: both fixes compact the child output and give empty TDigests zero-length null rows. #24103 additionally covers nullable percentile values and validity-mask boundaries at 31/32/33/65/257 rows, avoids a second validity-count synchronization, checks compact child-size overflow, and keeps the bounded grid-stride launch with a compile-time dense/compact specialization. Its focused regressions and full local QUANTILES_TEST passed (388/388) on current main. I will defer to the maintainers on which implementation to retain and am happy to close #24103 or provide the incremental commits.

File: cpp/src/quantiles/tdigest/tdigest.cu [hunk 5]
```diff
@@ -207,8 +213,8 @@ std::unique_ptr<column> compute_approx_percentiles(tdigest_column_view const& in
 
   auto percentiles_cdv = column_device_view::create(percentiles, stream);
 
-  // leaf is a column of size input.size() * percentiles.size()
-  auto const num_output_values = input.size() * percentiles.size();
+  // Empty input digests have zero-length output ranges, so the leaf is compact.
+  auto const num_output_values = (input.size() - num_empty_tdigests) * percentiles.size();
```
Comment (wjxiz1992) [inline cpp/src/quantiles/tdigest/tdigest.cu]: [P1] Compute and validate the compact child size in 64 bits. Both operands here are cudf::size_type (int32_t), so the multiplication can overflow before make_fixed_width_column. The new kernel then computes the dense pair count in int64_t and grid-strides over every pair. For example, 65,537 non-empty rows times 65,537 percentiles wraps the allocated child size to 131,073, while the kernel processes 4,295,098,369 pairs and writes out of bounds. Please calculate the product as thread_index_type, reject values above numeric_limits<size_type>::max(), and only then cast to size_type.

**Review by wjxiz1992 — CHANGES_REQUESTED** [review]
Requesting one correctness fix. The compact child allocation still computes its size in 32 bits while the new grid-stride kernel intentionally processes a 64-bit pair count. That mismatch can turn an unsupported oversized output into out-of-bounds GPU writes. Otherwise the mixed-empty compact-list fix looks sound. The C++ build/test matrix is green; the two devcontainer failures are image-pull infrastructure failures, and pr-builder only aggregates them.

**Review by bdice — APPROVED** [review]
